assess/utility/objectcache.py

```diff
@@ -0,0 +1,159 @@
+        if index is not None:
+            return self._object_cache[key][index]
+        return None
+
+    def remove_data(self, data=None, key=None, key_function=lambda data: data.pid):
+        """
+        Remove data object from cache. Returns `True` if the object was removed,
+        otherwise `False`.
+
+        :param data: data object to be removed
+        :param key: key where data is stored
+        :param key_function: function to get the key from data
+        :return: `True` if removal was successful, `False` otherwise
+        """
+        if key is None:
+            key = key_function(data)
+        try:
+            data_array = self._object_cache[key]
+            data_array.remove(data)
+            if len(self._object_cache[key]) == 0:
+                del self._object_cache[key]
+            return True
+        except ValueError:
+            pass
+        except KeyError:
+            pass
+        return False
+
+    def data_index(self, value=None, key=None, remember_error=False,
+                   validate_range=False, value_function=lambda data: data.tme,
+                   range_end_value_function=lambda data: data.exit_tme):
+        """
+        Returns index of closest value specified by *key* and *value*.
+
+        :param value: value to look for
+        :param key: key to look for
+        :param remember_error: remember mismatched keys, defaults to `False`
+        :param validate_range: check if value is in valid range for closest value
+        :param value_function: function to get the value for comparison from data
+        :param range_end_value_function: function to get end value for range
+            comparison from data
+        :return: closest data object
+        """
+        try:
+            data_array = self._object_cache[key]
+            value_array = [value_function(node) for node in data_array]
+            index = bisect.bisect_right(value_array, value) - 1
+        except KeyError:
+            if remember_error:
+                self.faulty_nodes.add(key)
+                logging.getLogger(self.__class__.__name__).info(
+                    "error for %s (%d)", key, value)
+            raise DataNotInCacheException(key, value)
+        else:
+            if validate_range:
+                selected_object = self._object_cache[key][index]
+                start_value = value_function(selected_object)
+                end_value = range_end_value_function(selected_object)
+                if value < start_value or value > end_value:
+                    if remember_error:
+                        self.faulty_nodes.add(key)
+                        logging.getLogger(self.__class__.__name__).info(
+                            "error for %s (%d)", key, value)
+                    raise DataNotInCacheException(key, value)
+
+            return index
+
+    def clear(self):
+        """
+        Clear the current state of the cache.
+        """
+        del self._object_cache
```

Comment:
You can drop the ``del ...`` altogether. Just replacing the attribute has the same effect.